README.md
Abelian(Direct) Connector
=========

API for sending/receiving Direct messages using Abelian

Current version of API allows to receive and send Direct messages using Abelian Direct implementation.

Abelian should be installed separately and added to `config.js`.

Send API endpoint is `/api/v1/send` and takes `to, subject, message` parameters and  file attachment 'file' over HTTP POST.

Receiver (listener) is continiously check Abelian for incoming messages and submit them into DRE ingestion API


##Quick up and running quide

###Prerequisites

- Node.js (v0.10+) and NPM
- Grunt.js

```
# you need Node.js and Grunt.js installed

# Usage (install)
npm install
node server.js
```

##Test:

```
npm install mocha -g
mocha test/test-send.js -R spec
```


## Contributing

Contributors are welcome. See issues https://github.com/amida-tech/abelian-connector/issues

## Release Notes

See release notes [here] (./RELEASENOTES.md)

## License

Licensed under [Apache 2.0](./LICENSE)

RELEASENOTES.md
# Abelian (DIRECT) Connector Release Notes

# v1.0.0 - March 16, 2015

This is initial release of the project

- Ability to receive messages from Abelian and ingest them into DRE
- API to send Direct messages using Abelian


test/test-send.js

var expect = require('chai').expect;
var supertest = require('supertest');
var api = supertest.agent('http://localhost:3001');
var fs = require('fs');
var path = require('path');

describe('Send API', function() {
    var sampleFile = '';


    it('Send Endpoint POST with CCDA', function(done) {
        var filepath = path.join(__dirname, './artefacts/CCD-sample.xml');
        api.post('/api/v1/send')
            .attach('file', filepath)
            .field('to','patient@direct.amida-demo.com')
            .field('subject', 'your health record')
            .field('message', 'please find your latest data attached')
            .expect(200)
            .end(function(err, res) {
                if (err) {
                    return done(err);
                } else {
                    expect(res.body).to.deep.equal({});
                    done();
                }
            });
    });
});

lib/inbox.js
var request = require('request');
var async = require('async');
var MailParser = require("mailparser").MailParser;
var _ = require('underscore');

function parseMail(email, callback) {

    var mailparser = new MailParser({
        unescapeSMTP: true
    });
    mailparser.on("end", function(mail_object) {

        //Throw errors if required fields aren't in object.

        if (_.isUndefined(mail_object.from) || _.isUndefined(mail_object.to) || _.isUndefined(mail_object.date) || _.isUndefined(mail_object.messageId)) {
            callback('Malformed Mail Object is Unparseable.');
        } else {

            var parsed_object = {
                sender: mail_object.from[0].address,
                recipient: mail_object.to[0].address,
                date: mail_object.date,
                message_id: mail_object.messageId,
                user: mail_object.to[0].address.split("@")[0]
            };

            //Flag MDNs via content-type evaluation.
            var content_type_split = mail_object.headers['content-type'].split(";");
            for (var i in content_type_split) {
                content_type_split[i] = content_type_split[i].trim();
            }

            parsed_object.mdn_flag = false;

            if (_.contains(content_type_split, 'multipart/report') === true) {
                if (_.contains(content_type_split, 'report-type=disposition-notification') === true) {
                    parsed_object.mdn_flag = true;
                }
            }

            if (mail_object.subject) {
                parsed_object.subject = mail_object.subject;

                //flag Abelian MDNs
                if (mail_object.subject.indexOf("Processed") >= 0) {
                    parsed_object.mdn_flag = true;
                }
            }
            if (mail_object.text) {
                parsed_object.contents = mail_object.text;
            }

            if (mail_object.attachments) {
                parsed_object.attachments = mail_object.attachments;
            }

            //clean mdns (attachments)
            if (parsed_object.mdn_flag) {
                delete parsed_object.attachments;
            }

            callback(null, parsed_object);

        }

    });
    mailparser.write(email);
    mailparser.end();
};

//calls API for single message and un-mimes it
function message(entry, callback) {
    var endpoint = entry.id;
    console.log("getting message ", endpoint);

    var options = {
        url: endpoint,
        method: 'GET',
    };

    request(options, function(err, response) {
        if (err) {
            console.error("error ", err);
            callback(err);
        } else {
            //got message
            //TODO: unmime it
            console.log(response.body.substring(0, 100));
            var text = response.body;


            var m = parseMail(text, function(err, data) {
                if (entry.to_delete) {
                    console.log("deleting email ", endpoint);

                    delete_message(endpoint, function(err) {   
                        console.log('err', err);
                        callback(null, data);
                    });

                } else {
                    callback(null, data);
                }

            });
        }
    });
}

function delete_message(endpoint, callback) {
    //actual call here
    var options = {
        url: endpoint,
        method: 'DELETE',
    };

    request(options, function(err, response) {
        if (err) {
            console.error("error ", err);
            callback(err);
        } else {
            //got inbox index
            console.log("message ", endpoint, " deleted");
            callback(null, response);
        }
    });

}

function messages(endpoint, index, to_delete, callback) {
    var calls = [];

    console.log("index.entry ", index["entry"]);
    for (var i = 0; i < index["entry"].length; i++) {
        console.log("flag entry ", index["entry"][i].id, "for deletion");
        index["entry"][i].to_delete = to_delete;
    }

    async.map(index["entry"], message, function(err, results) {
        if (err) {
            callback(err);
        } else {
            console.log()
            callback(null, results);
        }
    });
}

function inbox(endpoint, to_delete, callback) {
    console.log("getting inbox and delete messages flag=", to_delete);

    //actual call here
    var options = {
        url: endpoint + "/Messages",
        method: 'GET',
    };

    request(options, function(err, response) {
        if (err) {
            console.error("error ", err);
            callback(err);
        } else {
            //got inbox index
            var index = JSON.parse(response.body);

            console.log("index ", index);

            messages(endpoint, index, to_delete, function(err, response) {
                if (err) {
                    console.error('error ', err);
                    callback(err);
                } else {
                    callback(null, response);
                }
            })

        }
    });

}

module.exports = inbox;


lib/package.js
var moment = require('moment');
var MailComposer = require("mailcomposer").MailComposer;
var uuid = require('node-uuid');

// gets email param with to/from/subject/message
// gets array of attachments like {'fileName':'blah', 'contents':'content goes here'}
function package(email, attachments, callback) {

    //timestamp with now
    var formatted_date = moment().format("ddd, DD MMM YYYY HH:mm:ss ZZ");

    var origin = email.from.split("@")[1];
    var message_id = uuid.v4() + "@" + origin;    

    var mailcomposer = new MailComposer();

    //process all attachments
    /*
        attachments=[
            {'fileName':'sample.xml', 'contents':'raw text content'}
        ]
    */
    if (attachments) {
        //console.log('attachments ', attachments);
        for (var i=0; i<attachments.length; i++) {
            console.log("processing attachemt ", attachments[i].fileName);
            mailcomposer.addAttachment(attachments[i]);
        }
    }

    //process envelope
    mailcomposer.setMessageOption({
        from: email.from,
        to: email.to,
        subject: email.subject,
        body: email.body
    });
    mailcomposer.addHeader("date", formatted_date);
    mailcomposer.addHeader("Message-ID", message_id);
    mailcomposer.buildMessage(function(err, packagedMessage) {
        if (err) {
            callback(err);
        } else {
            callback(null, packagedMessage);
        }
    });
}

module.exports = package;


examples/amida-env.js
var fs = require('fs');
var pack = require('../lib/package.js');
var send = require('../lib/send.js');

var phr = "http://phr.amida-demo.com:8085";
var ehr = "http://ehr.amida-demo.com:8085";

var hl7 = fs.readFileSync('../test/artefacts/CCD-sample.xml', "utf8").toString();


//email from patient to doc with attachment (hl7)
pack({
    'from': 'john@direct.phr.amida-demo.com',
    'to': 'doctor@direct.ehr.amida-demo.com',
    'subject': 'My latest record',
    'body': 'See attached. -John'
}, [{
    'fileName': 'hl7.xml',
    'contents': hl7
}], function(err, mime) {
    console.log("error: ",err);
    console.log("envelope: ", mime.substring(0,750));

    send(phr, mime, function(err) {
        if (err) {
            console.log("error: ", err);
        } else {
            console.log("message sent");
        }
    })

});

server.js
var config = require('./config.js');

console.log("config: ", config);

var pack = require('./lib/package.js');
var send = require('./lib/send.js');
var inbox = require('./lib/inbox.js');

var express = require('express');
var path = require('path');
var logger = require('morgan');
var bodyParser = require('body-parser');

var http = require('http');
var FormData = require('form-data');

var fs = require('fs');

var multer = require('multer');
var autoReap = require('multer-autoreap');
autoReap.options = {
    reapOnError: true
};


var async = require('async');

var app = express();

app.all('*', function(req, res, next) {
    console.log("all request");
    res.header('Access-Control-Allow-Origin', '*');
    res.header('Access-Control-Allow-Methods', 'GET,POST,PUT,HEAD,DELETE,OPTIONS');
    res.header('Access-Control-Allow-Headers', 'Content-Type,X-Requested-With');
    next();
});

app.use(multer({
    dest: './uploads/'
}))
app.use(autoReap);

app.use(logger('dev'));

app.use('/docs', express.static('./swagger'));

app.get('/interop', function (req, res) {
    res.send("Direct Node Server is up and running!!!");
});

//send new message
app.post('/api/v1/send', function(req, res) {

    var filename = req.files.file.path;
    var data = fs.readFileSync(filename).toString();

    var to = req.body.to; 
    var from = req.body.from; //TODO: use patient's domain from config
    var subject = req.body.subject;
    var message = req.body.message;

    var memberId = "STAN01";

    //console.log(data);
    console.log("/api/v1/send");
    console.log("to: ", to);
    console.log("from: ", from);
    console.log("subject: ", subject);
    console.log("message: ", message);

    //package email
    var attachments = [{
        'filename': filename,
        'contents': data
    }];

    var email = {
        'from': from,
        'to': to,
        'subject': subject,
        'body': message
    };

    console.log("email: ", email);

    pack(email, attachments, function(err, mime) {
        if (err) {
            res.statusCode = 500;
            res.send(err);
        } else {
            //send email
            send(config.abelian, mime, function(err) {
                if (err) {
                    res.statusCode = 500;
                    res.send(err);

                } else {
                    console.log(mime.substring(0, 1000));
                    console.log("...skipping rest of mime envelope");
                    console.log("everything is ok");
                    res.send('OK');
                }
            });
        }
    });


});


//get list of messages
app.get('/api/v1/inbox', function(req, res) {
    //takes query param memberId /doctor e.g.  username/email address
    //var memberId = req.query.memberId || "STAN01";

    inbox(config.abelian, false, function(err, response) {
        if (err) {
            res.statusCode = 500;
            res.send(err);

        } else {
            console.log(response);
            console.log("everything is ok");
            res.send(response);
        }
    })

    //res.status(200).end();
});



var server = app.listen(3001, function() {

    var host = server.address().address
    var port = server.address().port

    console.log('Listening at http://%s:%s', host, port)

    var mailListener = require('./lib/mail-listener');

    function timerMail() {
        mailListener(function(err, results) {
            if (err) {
                console.error(err);
            } else {
                console.log("got mail  #", results.length);
                //TODO: process each email into DRE
                for (var j = 0; j < results.length; j++) {
                    console.log("processing email ", results[j]);
                    //console.log(results[j].attachments[0].content.toString());

                    //DRE Ingest call here
                    console.log("posting to ingest api: ", config['dre-ingest-api'] + '/api/v1/ingest');
                    var patKey = results[j].user;

                    if (results[j].attachments) {
                        var form = new FormData();
                        form.append('file', results[j].attachments[0].content, {filename: 'bluebutton.xml'});
                        form.append('patKey', patKey);
                        form.submit(config['dre-ingest-api'] + '/api/v1/ingest', function(err, res) {
                            //res.resume(); // for node-0.10.x
                        });
                    }
                }
            }
        });
    }

    //mail listener (every 30 seconds)
    console.log("listening for mail every 15 seconds");
    setInterval(timerMail, 15000);

})
